### DISCUSSION TITLE
Is there any way to debug HMR and/or improve Nuxt start / refresh times?

### DISCUSSION THREAD

**Discussion by matthew-dean** [nuxt/nuxt#23533] [Questions]
I'm working with a scenario where the latest Nuxt refuses to ever hot-reload a module. Is there a way to either:

1. Log all globs / files that Nuxt is watching for HMR?
2. Log if HMR is working / not working?
3. Adjust the files / definitions that Nuxt is watching for HMR?
4. "Soft reload" or hit some key after Nuxt is loaded so that it can poll / reload files if it needs to? Like any way to refresh without having to do a 30-second restart per file change?

**Comment by matthew-dean** [comment]
Can anyone help at all? It's really hard to use Nuxt without HMR, as any change requires restarting the Nuxt server, which takes about 30 seconds.

**Reply by danielroe** [reply]
What are you changing that isn't ending up with HMR?

**Reply by matthew-dean** [reply]
@danielroe I'm not sure what your question means? Any file (source `.ts` and `.vue` files mostly, but also lots of `.js` files) I edit never results in that module being refreshed in the browser... Well, I shouldn't say never. Once in a great while, I see the console say that something was hot reloaded after I edit / save. (Maybe usually a `.vue` file representing a page?) But that's very very rare. I've been setting up a Nuxt site for over a month, it and it rarely, rarely gives this message, and most times I have to restart the Nuxt server after every file change, which takes up to a minute to get back up and running again.

So, is there any way to debug this? To figure out what's going on?

**Reply by matthew-dean** [reply]
Are there known paths where HMR will not be set up? Like if there are any existing module resolution errors (but not with paths that are currently loaded), could that be an issue?

**Reply by matthew-dean** [reply]
@danielroe Alternatively, is there any way to start my Nuxt server in under 60 seconds to make the lack of HMR in my case (in the meantime) less painful? Like, is there a setting for incremental builds that can check file stats for modules to decide whether to rebuild or not?

**Reply by danielroe** [reply]
That sounds like a bug. Can you provide a reproduction?

**Reply by matthew-dean** [reply]
> Can you provide a reproduction?

I can't. It's a very large project of proprietary code. Which is why this question is around what debugging tools / settings / configs are available to debug HMR. And, if there are none, what tools / configs are available (like near-instant incremental builds) so that the _lack_ of HMR is less of a problem. Both Vite and TypeScript have incremental build capabilities, so I'm not sure why Nuxt is crawling the entire module tree each time?

**Reply by matthew-dean** [reply]
@danielroe For perspective, the project is 139 MB of TS, JS, and Vue (about 1,500 files), so I'm not entirely surprised that Nuxt or one of the libraries it depends on is choking and failing; I'm more curious about how to investigate _when_ it's choking and why, and what are some possible workarounds.

**Reply by danielroe** [reply]
I would debug an issue with HMR with a custom vite plugin.

Nuxt doesn't take over your HMR behaviour but it does watch changes across your project.

If nothing is changing (and no hmr log is printed) then the issue could be with the vite watcher or the Nuxt watcher. And it varies on your OS.

You can try changing the Nuxt watcher with an experimental option we introduced recently. I know vite is considering using the parcel watcher in future.

**Reply by matthew-dean** [reply]
@danielroe The `parcel` option didn't change the behavior but I appreciate the help! I'll try a Vite plugin.

**Reply by matthew-dean** [reply]
@danielroe Okay, I fixed it! Thanks for the tips. `useFsEvents` was failing entirely for chokidar. When I switched it to `usePolling` it works fine.

**Reply by danielroe** [reply]
Delighted to hear it. So it was an issue with the vite watcher, it sounds like, which is chokidar?

**Reply by matthew-dean** [reply]
@danielroe I assume so? Since there's not a way to debug that I could find, it seems that something wasn't informing the Nuxt process?

**Comment by bmulholland** [comment]
Hey, same question. How do I debug this? For me, HMR isn't working for a fresh upgrade. I have to restart the server for changes. The console shows "page reload" and the file that changed, so the watcher is working, but the browser never updates.

**Reply by bmulholland** [reply]
Even refreshing the page doesn't work -- the contents are stale until I restart the server. I think that rules out browser-specific issues?

**Comment by haayman-imagem** [comment]
Same question, different issue. I keep getting CONNECTION_REFUSED messages lately. Tried to change the protocol and port. When I set the protocol to http and e.g. port 5000 I do see a process listening to that port in netstat, but I still get the CONNECTION_REFUSED message. So I really would like vite to be more verbose about what's going on
